src/workerd/io/trace.c++

```diff
@@ -714,6 +806,711 @@ tracing::Exception::Exception(rpc::Trace::Exception::Reader reader)
+        jsrpc.copyTo(infoBuilder.initJsrpc());
+      }
+      KJ_CASE_ONEOF(custom, tracing::CustomInfo) {
+        auto customBuilder = infoBuilder.initCustom(custom.size());
+        for (size_t n = 0; n < custom.size(); n++) {
+          custom[n].copyTo(customBuilder[n]);
+        }
+      }
+    }
+  }
+}
+
+tracing::SpanOpen tracing::SpanOpen::clone() {
+  constexpr auto cloneInfo = [](kj::Maybe<Info>& info) -> kj::Maybe<tracing::SpanOpen::Info> {
+    return info.map([](Info& info) -> tracing::SpanOpen::Info {
+      KJ_SWITCH_ONEOF(info) {
+        KJ_CASE_ONEOF(fetch, tracing::FetchEventInfo) {
+          return fetch.clone();
+        }
+        KJ_CASE_ONEOF(jsrpc, tracing::JsRpcEventInfo) {
+          return jsrpc.clone();
+        }
+        KJ_CASE_ONEOF(custom, tracing::CustomInfo) {
+          kj::Vector<tracing::Attribute> attrs(custom.size());
+          for (size_t n = 0; n < custom.size(); n++) {
+            attrs.add(custom[n].clone());
+          }
+          return attrs.releaseAsArray();
+        }
+      }
+      KJ_UNREACHABLE;
+    });
+  };
+  return SpanOpen(operationName.map([](auto& str) { return kj::str(str); }), cloneInfo(info));
+}
+
+tracing::SpanClose::SpanClose(EventOutcome outcome): outcome(outcome) {}
+
+tracing::SpanClose::SpanClose(rpc::Trace::SpanClose::Reader reader): outcome(reader.getOutcome()) {}
+
+void tracing::SpanClose::copyTo(rpc::Trace::SpanClose::Builder builder) {
+  builder.setOutcome(outcome);
+}
+
+tracing::SpanClose tracing::SpanClose::clone() {
+  return SpanClose(outcome);
+}
+
+namespace {
+tracing::Onset::Info getInfoFromReader(const rpc::Trace::Onset::Reader& reader) {
+  auto info = reader.getInfo();
+  switch (info.which()) {
+    case rpc::Trace::Onset::Info::FETCH: {
+      return tracing::FetchEventInfo(info.getFetch());
+    }
+    case rpc::Trace::Onset::Info::JSRPC: {
+      return tracing::JsRpcEventInfo(info.getJsrpc());
+    }
+    case rpc::Trace::Onset::Info::SCHEDULED: {
+      return tracing::ScheduledEventInfo(info.getScheduled());
+    }
+    case rpc::Trace::Onset::Info::ALARM: {
+      return tracing::AlarmEventInfo(info.getAlarm());
+    }
+    case rpc::Trace::Onset::Info::QUEUE: {
+      return tracing::QueueEventInfo(info.getQueue());
+    }
+    case rpc::Trace::Onset::Info::EMAIL: {
+      return tracing::EmailEventInfo(info.getEmail());
+    }
+    case rpc::Trace::Onset::Info::TRACE: {
+      return tracing::TraceEventInfo(info.getTrace());
+    }
+    case rpc::Trace::Onset::Info::HIBERNATABLE_WEB_SOCKET: {
+      return tracing::HibernatableWebSocketEventInfo(info.getHibernatableWebSocket());
+    }
+    case rpc::Trace::Onset::Info::RESUME: {
+      return tracing::Resume(info.getResume());
+    }
+    case rpc::Trace::Onset::Info::CUSTOM: {
+      return tracing::CustomEventInfo();
+    }
+  }
+  KJ_UNREACHABLE;
+}
+kj::Maybe<kj::String> getScriptNameFromReader(const rpc::Trace::Onset::Reader& reader) {
```

Comment:
almost all getters have missing const qualifiers. can you add them before landing?

Reply:
These aren't getters. They are statics.

Reply:
It's hard to distinguish when the static function names doesn't start with capital letter (like node.js does)...